### ISSUE TITLE
Code review complete - discussion of project roadmap

### ISSUE THREAD

**Issue by lvotapka** [anandojha/qmmmrebind#7]
I've completed a pretty detailed review of this project's code, so I wanted to open an issue thread to discuss and lay out our project "roadmap" - plans and priorities we will lay out in the form of Github "milestones".

First of all, it's worth saying, @anandojha, that this is a very impressive body of code, you have certainly done the lion's share of good work here with the code and the documentation, so I'm confident that with some minor, but meaningful, adjustments to the code, we can:
A) Find and fix the bugs preventing QMMMRebind from performing its best on all current systems, and,
B) Prepare the project for publication and deployment to the scientific community.

In my review of the code, I discovered the following issues for us to discuss (any of these can be moved to different issues for more detailed discussion, if needed). These issues are ranked in order of relative importance/urgency

1. Parametrize.py needs to be broken up into smaller modules to improve the ease of coding and testing, and also to avoid software entropy (the code becoming too complex to expand). I will likely open a separate issue to discuss how best to do this.
2. The DRY (don't repeat yourself) paradigm needs to be followed more strictly. Any time two codes in separate places resemble each other, they should probably be combined into a single function. This will decrease line counts, reduce software entropy, and make testing and bugfixing much easier. (Examples: a single function to read Gaussian log files, a function to read other files that returns lines based on which section is desired, etc.)
3. Class inheritance should be used. For instance, the classes PrepareGaussianGuest and PrepareGaussianHostGuest share very similar parameters and methods, therefore they should have a superclass that each of them inherit from, to reduce line count and entropy, and to make testing and expansion easier.
4. The program needs to be able to support a host QM region that is not necessarily contiguous - the residues should not have to be directly bonded in a single line to be included in the QM region. This is because the guest will be surrounded on all sides by host residues that may be tens or hundreds of amino acids away in the sequence, but providing important interactions to the guest.
5. The addition of terminal atoms on the host residues should be automated, perhaps with pdb4amber. Care must be taken to ensure that the proper parameters will be placed on the proper atoms, since the host molecule is being modified from its original structure.
6. We should contemplate passing a persistent object between the classes of QMMMRebind (similar to the Model() object in SEEKR2), which contains, among other things, atom counts and identities for the host and guest molecules. This will allow us to perform easy checks and tests at each stage of the calculation to ensure that the parameters are being passed along properly.
7. The adaptation of the modified Seminario method for extracting bond and angle parameters from the Hessian should be carefully tested.
8. Generation of input files for GAMESS as an alternative to Gaussian - class inheritance will help with this immensely.
9. Try not to do "from library import *", but rather do "import library" and use functions using the full library reference.
10. Sort all output/input files into directories based on the calculation stages. Also, these files should use clearer naming.
11. Some variables and attributes are not given correct initialization (Example: a variable supposed to be an int is initialized as "").
12. When converting from XML to AMBER, or vice versa, parameters should be checked atom-by-atom (or parameter by parameter)
13. Use DistUtils to make a qmmmrebind package so that string paths aren't necessary within run scripts.

These are the improvements that I think should be made. Once you share your thoughts about these ideas, @anandojha, we can open up milestones for them and partition the work assignments.

**Comment by anandojha** [comment]
Thank you Lane for the review. This is awesome!!  I think we can go in two directions in parallel, i.e. code improvement and test cases for the purpose of the manuscript. We can also decide on the test cases if we may want to change them and use a JAK2 vs JAK3 inhibitor (which can be done easily). Let us meet and discuss and maybe, you can help me in taking some load off by pushing some of the changes you think are manageable from your side.

**Comment by lvotapka** [comment]
Alright let's meet very soon about this. And yes, I have every intention of "rolling up my sleeves" and working on some of this code myself.

**Comment by anandojha** [comment]
Thank you !!
